src/sentry/integrations/source_code_management/repo_trees.py

```diff
@@ -0,0 +1,310 @@
+from __future__ import annotations
+
+import logging
+from abc import ABC, abstractmethod
+from collections.abc import Sequence
+from typing import Any, NamedTuple
+
+from sentry.integrations.models.organization_integration import OrganizationIntegration
+from sentry.shared_integrations.client.base import BaseApiResponseX
+from sentry.shared_integrations.exceptions import ApiError, IntegrationError
+from sentry.utils.cache import cache
+
+logger = logging.getLogger(__name__)
+
+
+# Read this to learn about file extensions for different languages
+# https://github.com/github/linguist/blob/master/lib/linguist/languages.yml
+# We only care about the ones that would show up in stacktraces after symbolication
+SUPPORTED_EXTENSIONS = ["js", "jsx", "tsx", "ts", "mjs", "py", "rb", "rake", "php", "go", "cs"]
+EXCLUDED_EXTENSIONS = ["spec.jsx"]
+EXCLUDED_PATHS = ["tests/"]
+
+
+class RepoAndBranch(NamedTuple):
+    name: str
+    branch: str
+
+
+class RepoTree(NamedTuple):
+    repo: RepoAndBranch
+    files: list[str]
+
+
+# Tasks which hit the API multiple connection errors should give up.
+MAX_CONNECTION_ERRORS = 10
+
+# When the number of remaining API requests is less than this value, it will
+# fall back to the cache.
+MINIMUM_REQUESTS_REMAINING = 200
+
+
+class RepoTreesIntegration(ABC):
+    """
+    Base class for integrations that can get trees for an organization's repositories.
+    It is used for finding files in repositories and deriving code mappings.
+    """
+
+    CACHE_SECONDS = 3600 * 24
+
+    @abstractmethod
+    def get_client(self) -> RepoTreesClient:
+        """Returns the client for the integration. The client must be a subclass of RepositoryClient."""
+        raise NotImplementedError
+
+    @property
+    def org_integration(self) -> OrganizationIntegration | None:
+        raise NotImplementedError
+
+    @property
+    def org_slug(self) -> str:
+        raise NotImplementedError
+
+    def get_trees_for_org(self) -> dict[str, RepoTree]:
+        trees = {}
+        if not self.org_integration:
+            raise IntegrationError("Organization Integration does not exist")
+        repositories = self._populate_repositories()
+        if not repositories:
+            logger.warning("Fetching repositories returned an empty list.")
+        else:
+            trees = self._populate_trees(repositories)
+
+        return trees
+
+    def _populate_repositories(self) -> list[RepoAndBranch]:
+        cache_key = f"githubtrees:repositories:{self.org_slug}"
+        repositories: list[RepoAndBranch] = cache.get(cache_key, [])
+
+        if not repositories:
+            # XXX: Switch to the integration get_repositories to be more generic
+            # Remove unnecessary fields from GitHub's API response
+            repositories = [
+                RepoAndBranch(name=repo["full_name"], branch=repo["default_branch"])
+                for repo in self.get_client().get_repositories(fetch_max_pages=True)
```

Comment:
In the follow-up pull request, I will switch to using `self.get_repositories()` but I need to integrate the two functions first.

Reply:
I handled this in the current PR.